clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -221,8 +222,45 @@ public int partitionLeaderEpoch() {
         return buffer.getInt(PARTITION_LEADER_EPOCH_OFFSET);
     }
 
+    public boolean validateBatchSize() {
```

Comment:
@hachikuji Underflow case is not difficult. But the problem with the overflow is that `hasNext` has the condition `readRecords < numRecords` so the case for overflow in `next()` which is `readRecords >= numRecords` will never be checked. What could be done is to force call on next again after going through the whole recordset (without checking `hasNext`) to see if this triggers exception... (good) or doesn't (bad). This also could be a separate function. WDYT?